### ISSUE TITLE
Timezone issue with timestamp columns if server doesn't use UTC.

### ISSUE THREAD

**Issue by dveeden** [github/gh-ost#161]
After running a gh-ost with -test-on-replica I found differences in two timestamp columns.

I found this by doing a 
```
select 
  * 
from 
  t1 a 
  inner join _t1_gho b on a.id=b.id 
where 
  a.c1<>b.c1 
  or a.c2<>b.c2
  ...
```

Record 1
```
   timestamp_notified: 2016-08-17 13:08:04
   timestamp_notified: 2016-08-17 11:08:04
```

Record 2
```
   timestamp_notified: 2016-08-17 13:08:19
  timestamp_finalized: 2016-02-01 23:39:30
   timestamp_notified: 2016-08-17 11:08:19
  timestamp_finalized: 2016-02-01 22:39:30
```

From the definition:
```
  `timestamp_notified` timestamp NULL DEFAULT NULL,
  `timestamp_exported` timestamp NULL DEFAULT NULL,
  `timestamp_finalized` timestamp NULL DEFAULT NULL,
```

**Comment by dveeden** [comment]
Environment: CentOS 6, MySQL 5.6.27
/etc/localtime is a link to /usr/share/zoneinfo/Europe/Amsterdam

@@global.time_zone is set to SYSTEM.

**Comment by shlomi-noach** [comment]
Are these all the rows of the table? How many rows does the table have?
Also, to speed up the investigation, would you guess these rows were inserted/added/manipulated throughout the migration process?

**Comment by shlomi-noach** [comment]
Also, please confirm that on `record 2` there are:

- one TZ difference of 2 hours
- one TZ difference of 1 hour

-- in the same record. Please double check this?

**Comment by dveeden** [comment]
Rows with differences: 7085
Total number of rows (from show table status): 31982317

Yes I do think these were modified while gh-ost was running.

**Comment by dveeden** [comment]
Yes I verified the 2h/1h difference... I have no clue about why that is..

**Comment by dveeden** [comment]
This is with gh-ost from the v1.0.8 tag/release.

**Comment by shlomi-noach** [comment]
was `gh-ost` executing from the very same replica host, or from another host? If from another host, can you check what the TZ is on said host?

**Comment by dveeden** [comment]
The very same replica host.

**Comment by dveeden** [comment]
Tested with v1.0.9 but hit #157 so used a build from today's master. It shows the same issue.

```
*************************** 1. row ***************************
                id: 29793038
Original Timestamp: 2016-08-18 10:01:56
  gh-ost Timestamp: 2016-08-18 08:01:56
```

It looks like the timestamp was converted from CEST to UTC and then inserted.

Observed from the binlogs:
* The row is inserted with NULL values for the timestamp and gets at least one update which doesn't touch the timestamp columns. This is while gh-ost runs.
* Then the row gets updated with a timestamp

From the mysqlbinlog output:
```
### UPDATE `t1`.`foo`
### WHERE
###   @1=29793038
…
### SET
###   @21=1471507316

### UPDATE `t1`.`_foo_gho`
### WHERE
###   @1=29793038
…
### SET
###   @21=1471500116
```

**Comment by shlomi-noach** [comment]
@dveeden thank you for all this input. Also, last question, though I suspect I already know the answer: master and replicas are all in the same timezone, correct?

And last-last question, is master in `STATEMENT` based replication?

**Comment by dveeden** [comment]
With `SET GLOBAL time_zone='+00:00'` set I couldn't reproduce the issue.
The 1 and 2 hour offsets could have been the 1 or 2 hour difference between UTC and CEST/CET

Possible solutions:
1. do a `SET SESSION time_zone='+00:00'` on the connection which applies events
2. Convert the timestamp to @@session.time_zone before applying it. Don't use CONVERT_TZ() for that as not having the timezone database loaded is fairly common. So convert it in go.

**Comment by dveeden** [comment]
master and server use the same timezone.
master uses minimal RBR.

**Comment by shlomi-noach** [comment]
Will add fix once https://github.com/github/gh-ost/pull/172 is tested (sent to testing for the duration of the day)

**Comment by shlomi-noach** [comment]
meanwhile @dveeden you can:

```
git fetch origin
git checkout transactional-apply-dml -b tz-trx-fix
``` 

and proceed from there to assign the `time_zone` from within the trx and before the query is executed: https://github.com/github/gh-ost/pull/172/files#diff-55e7b291008bc31a7fac46abdad84feaR874

**Comment by dveeden** [comment]
@shlomi-noach yes planning to test this today

**Comment by shlomi-noach** [comment]
To clarify, there's now a dedicated branch: `apply-dml-tz` that should solve your issue, you only need to check it out.

**Comment by dveeden** [comment]
I did test it and it does solve this issue.  :+1:

**Comment by shlomi-noach** [comment]
👍 thanks for updating; Will merge come Monday

**Comment by shlomi-noach** [comment]
Fixed by #175

**Comment by shlomi-noach** [comment]
#175 doesn't actually fix this. #182 does, and is not merged yet. Repoening.

**Comment by shlomi-noach** [comment]
@dveeden are you able to lend a hand in testing https://github.com/github/gh-ost/pull/265, which takes a different approach on time-zone handling? Can you please test it on your setup?
Insofar it passes all tests on my side (unit + prod) and on another user side.

**Comment by shlomi-noach** [comment]
Further discussion in https://github.com/github/gh-ost/issues/247#issuecomment-253151080
